core/src/main/scala/org/http4s/QueryParam.scala

```diff
@@ -41,43 +42,82 @@ object QueryParamKeyLike {
  * Type class defining how to encode a `T` as a [[QueryParameterValue]]s
  * @see QueryParamCodecLaws
  */
-trait QueryParamEncoder[T] {
+trait QueryParamEncoder[T] { outer =>
   def encode(value: T): QueryParameterValue
+
+  /** QueryParamEncoder is a contravariant functor. */
+  def contramap[U](f: U => T): QueryParamEncoder[U] =
+    new QueryParamEncoder[U] {
+      override def encode(value: U) =
+        outer.encode(f(value))
+    }
+
 }
 
 object QueryParamEncoder {
 
   /** summon an implicit [[QueryParamEncoder]] */
   def apply[T](implicit ev: QueryParamEncoder[T]): QueryParamEncoder[T] = ev
 
-  def encodeBy[T, U: QueryParamEncoder](f: T => U): QueryParamEncoder[T] = new QueryParamEncoder[T] {
-    def encode(value: T): QueryParameterValue =
-      QueryParamEncoder[U].encode(f(value))
-  }
+  /** QueryParamEncoder is a contravariant functor. */
+  implicit val ContravariantQueryParamEncoder: Contravariant[QueryParamEncoder] =
+    new Contravariant[QueryParamEncoder] {
+      override def contramap[A, B](fa: QueryParamEncoder[A])(f: B => A) =
+        fa.contramap(f)
+    }
+
+  @deprecated("Use QueryParamEncoder[U].contramap(f)", "0.17")
```

Comment:
If we target this at release-0.16 branch, it'll be more useful for you.  And then update these deprecations accordingly.